types/jest/index.d.ts

```diff
@@ -73,6 +73,7 @@ type ExtractEachCallbackArgs<T extends ReadonlyArray<any>> = {
 ];
 
 declare namespace jest {
+    const mocked: typeof import('jest-mock').mocked;
```

Comment:
since we adding `jest-mock`  dependency here
should we need to  [import all types/functions](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L159to#L167) exported by `jest-mock` like this. No ?
so that it  looks consistent :)
or it doesn't matter

Reply:
For now it doesn't matter

Reply:
won't this break for people who have e.g.

```ts
function doSomething(jest.Mock<something, somethingElse>) {
}

doSomething(jest.mocked({}));
```

?

(Long term I want `@types/jest` to import all of its things from Jest itself (#44365), but as longs as that's not done, mixing seems less than ideal?)

Reply:
What you've provided isn't actually valid because `jest.Mock` is actually sort of poorly named `jest.MockedFunction`, where as `jest.mocked({})` returns `jest.MockedObject`.

However no it won't because the types should all be compatible.

<details>
<summary>here's some code showcasing this</summary>

```
import { mocked as tsJestMocked } from 'ts-jest/utils';
// import { jest } from '@jest/globals';
import * as MockTypes from 'jest-mock';
import * as jestMock from 'jest-mock';

// type MockTypes = jest;
// type MockTypes = typeof jestMock;
const mocked = jestMock.mocked;
// const mocked = jest.mocked;
// const mocked = tsJestMocked;

class Greeter {
  private readonly _greeting: string;

  public constructor(greeting: string) {
    this._greeting = greeting;
  }

  public get greeting() {
    return this._greeting;
  }

  public sayHello(target: string) {
    console.log(`${this._greeting} ${target}`);
  }
}

interface Point {
  x: string;
  y: number;
}

const makePoint = (x: string, y: number): Point => {
  return { x, y };
};

declare function doSomethingWithMockedFn(
  fn: jest.Mock<Point, [string, number]>
): void;
declare function doSomethingWithMockedMakePoint(
  fn: jest.Mocked<typeof makePoint>
): void;

declare function doSomethingWithMockedObject(fn: jest.Mocked<Point>): void;

declare function doSomethingWithMockedClass(
  fn: jest.MockedClass<typeof Greeter>
): void;

doSomethingWithMockedFn(mocked(makePoint));
doSomethingWithMockedMakePoint(mocked(makePoint));
doSomethingWithMockedObject(mocked({ x: '', y: 1 }));
doSomethingWithMockedClass(mocked(Greeter));
```
</details>

Regardless of what combination you use (e.g. types from `jest-mocked` + `mocked` from the `jest` global, everything from `jest-mocked`, everything from `ts-jest`, renaming the `jest.<type>` to be `MockTypes.<type>`, etc) the errors and passes are the same.

The first and last `doSomething` method calls have errors because of type issues (which basically boil down to the conditional typing of `mocked` resulting in a more accurate type than what's being used) while the middle two calls don't.

----

> Long term I want @types/jest to import all of its things from Jest itself (#44365), but as long as that's not the case, mixing seems less than ideal?

That was part of my goal here, as we're already mixing types with `jest-diff` and `pretty-format` - but I didn't want to change the existing types in this PR because of the size of that work.

However looking at it with fresh eyes, it'll be a better IDE experience anyway to go with #57776 so let's do that.


https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57776/files#r772003152

Reply:
(the problem with #57776 is it is mixing types too, but it's more subtle because they have slightly different names)